icu4c/source/common/localebuilder.cpp

```diff
@@ -0,0 +1,395 @@
+        while (U_SUCCESS(error) &&
+               nullptr != (key = iter->next(nullptr, error))) {
+            locale->setUnicodeKeywordValue(key, nullptr, error);
+        }
+    }
+}
+
+static void
+_setUnicodeExtensions(Locale* locale, const CharString& value, UErrorCode& error)
+{
+    // Add the unicode extensions to extensions_
+    CharString locale_str("und-u-", error);
+    locale_str.append(value, error);
+    _copyExtensions(
+        Locale::forLanguageTag(locale_str.data(), error),
+        locale, false, error);
+}
+
+LocaleBuilder& LocaleBuilder::setExtension(char key, StringPiece value)
+{
+    if (U_FAILURE(status_)) return *this;
+    if (!UPRV_ISALPHANUM(key)) {
+        status_ = U_ILLEGAL_ARGUMENT_ERROR;
+        return *this;
+    }
+    CharString value_str(value, status_);
+    char* p = value_str.data();
+    for (int32_t i = 0; i < value_str.length(); i++, p++) {
+        *p = (*p == '_') ? '-' : uprv_tolower(*p);
+    }
+    if (value_str.length() > 0 &&
+        !_isExtensionSubtags(key, value_str.data(), value_str.length())) {
+        status_ = U_ILLEGAL_ARGUMENT_ERROR;
+        return *this;
+    }
+    if (extensions_ == nullptr) {
+        extensions_ = new Locale();
+    }
+    if (uprv_tolower(key) != 'u') {
+        // for t, x and others extension.
+        extensions_->setKeywordValue(StringPiece(&key, 1), value_str.data(),
+                                     status_);
+        return *this;
+    }
+    _clearUAttributesAndKeyType(extensions_, status_);
+    if (value.length() > 0) _setUnicodeExtensions(extensions_, value_str, status_);
+    return *this;
+}
+
+LocaleBuilder& LocaleBuilder::setUnicodeLocaleKeyword(
+      StringPiece key, StringPiece type)
+{
+    if (U_FAILURE(status_)) return *this;
+    if (!ultag_isUnicodeLocaleKey(key.data(), key.length()) ||
+        (type.length() > 0 &&
+         !ultag_isUnicodeLocaleType(type.data(), type.length()))) {
+      status_ = U_ILLEGAL_ARGUMENT_ERROR;
+      return *this;
+    }
+    if (extensions_ == nullptr) extensions_ = new Locale();
+    extensions_->setUnicodeKeywordValue(key, type, status_);
+    return *this;
+}
+
+LocaleBuilder& LocaleBuilder::addUnicodeLocaleAttribute(
+    StringPiece value)
+{
+    CharString value_str(value, status_);
+    if (U_FAILURE(status_)) return *this;
+    char* p = value_str.data();
+    for (int32_t i = 0; i < value_str.length(); i++, p++) {
+        *p = uprv_tolower(*p);
+    }
+    if (!ultag_isUnicodeLocaleAttribute(value_str.data(), value_str.length())) {
+        status_ = U_ILLEGAL_ARGUMENT_ERROR;
+        return *this;
+    }
+    if (extensions_ == nullptr) {
+        extensions_ = new Locale();
+        extensions_->setKeywordValue(kAttributeKey, value_str.data(), status_);
+        return *this;
+    }
+
+    CharString attributes;
+    CharStringByteSink sink(&attributes);
+    UErrorCode get_error = U_ZERO_ERROR;
+    extensions_->getKeywordValue(kAttributeKey, sink, get_error);
```

Comment:
after this, if you do keep a separate UErrorCode, check for failure

Reply:
Actually, I don't need to. If there are no such attribute, then it will fail, but then the attributes will be empty, and therefore the while loop will just be skipped (since start == end) and therefore the whole thing will work in the end. But I could add a failure check (which mean there were no attribute in the extension so far, not really a real failure) and duplicate the code and return earlier.